### PR TITLE
1415 - Handle long interpreter directives in external programs

### LINKED ISSUE #1415: Molecule shouldn't call external programs directly to avoid shebang limitations
# Issue Type

- Bug report

# Molecule and Ansible details

```
$ ansible --version
ansible 2.6.2
```

```
$ molecule --version
molecule, version 2.16.0
```

Molecule installation method:

- pip

Ansible installation method:

- pip

# Desired Behavior

Molecule should work as expected even with virtual envs nested deeply in directory tree.

# Actual Behaviour

The length of shebangs in Linux is limited by `BINPRM_BUF_SIZE` set in `include/linux/binfmts.h`. By default `BINPRM_BUF_SIZE=128`. This value can't be changed without recompiling the kernel.

We currently run molecule with [Tox](https://github.com/tox-dev/tox) and Jenkins and sometimes, when path to some executable (i.e. ansible-playbook) becomes too long, Molecule fails. For example, here is an excerpt from Molecule's log running in our typical environment: 

```
--> Scenario: 'default'
--> Action: 'syntax'
<REDACTED>
DEBUG: COMMAND

/opt/jenkins/workspace/Molecule/test_role/roles/some_longer_than_usual_role_name/.tox/some_longer_than_usual_role_name-current/bin/ansible-playbook --skip-tags=molecule-notest,notest --diff --inventory=/tmp/longer_than_usual_role_name-current/ansible_inventory.yml --syntax-check /opt/jenkins/workspace/Molecule/test_role/roles/some_longer_than_usual_role_name/molecule/default/playbook.yml -vvv

ERROR: 

An error occurred during the test sequence action: 'syntax'. Cleaning up.
```

Executable `/opt/jenkins/workspace/Molecule/test_role/roles/some_longer_than_usual_role_name/.tox/some_longer_than_usual_role_name-current/bin/ansible-playbook` has rather long shebang:
```
#!/opt/jenkins/workspace/Molecule/test_role/roles/some_longer_than_usual_role_name/.tox/some_longer_than_usual_role_name-current/bin/python
```

Its length is 137 which obviously exceeds BINPRM_BUF_SIZE limitation mentioned above.

I see several possible **workarounds** for this problem:

* **Do not use tox and venvs.** This kills one of the most cool things about Molecule: the ability to test roles in different environments against multiple versions of Ansible.
* **Do not use long roles names and/or long tox env names.** We've been trying to do this for a while but it's not always possible and it's incredibly inconvenient.
* **Do not call external executables directly**, instead parse shebang (if said executable has one) and call the interpreter explicitly. Actually Tox took this exact approach some time ago:
    * https://github.com/tox-dev/tox/issues/66
    * https://github.com/tox-dev/tox/blob/46fca4c3c817f0a4728552e99e89ace9b2cc6234/tox/venv.py#L420

I've [tried implementing](https://github.com/metacloud/molecule/compare/master...kdask:shebang-limitations-workaround) this last workaround and it does seem to solve the problem. I don't know how I feel about this implementation though: looks a bit hacky to me, but, well :) Tell me what you think!

### PR DESCRIPTION
**Pull request by kdask**
Fixes #1415 

I'm not sure about how to test it, though. Any suggestions?

### CODE REVIEW

**Comment by retr0h** [comment]
fyi - this isn't something I have a good way of testing.  I'm not really inclined to merge this without a way to reproduce it through unit tests to verify it actually worked.  Also, the code submitted lowers the over all code coverage, since it also does not contain unit tests.

**Comment by retr0h** [comment]
I'm going to hold off from merging this.  As you know [Ansible is adopting Molecule](https://groups.google.com/forum/#!topic/ansible-project/ehrb6AEptzA), therefore I do not want to merge anything that may impact future architectural decisions, as I will no longer be leading the development of this project.

**Comment by gundalow** [comment]
Once the repo has been moved under Ansible we can look at getting this merged.

In the short term we welcome testing and feedback on this PR.

Thanks for your understanding.

**Comment by gundalow** [comment]
@kdask thank you for the PR. Looks like for this to progress we need to work out a way to test it, any ideas?

File: molecule/util.py [hunk 1]
```diff
@@ -95,6 +135,8 @@ def run_command(cmd, debug=False):
     :param debug: An optional bool to toggle debug output.
     :return: ``sh`` object
     """
+    if 'MOLECULE_LIMITED_SHEBANG' in os.environ:
```
Comment (webknjaz) [inline molecule/util.py]: maybe booleanize this?

`to_bool(os.getenv('MOLECULE_LIMITED_SHEBANG', False))`

**Comment by themr0c** [comment]
close/repopren to trigger travis build.

**Comment by ssbarnea** [comment]
Now that I look better a the code I realise that it is totally outside molecule scope to fix broken shebangs lines in scripts not produced by it.

Molecule is not the place to address this issue.

**Pull request closed without merging**